### PR TITLE
driver metadata: use addon keys as reference on system images

### PR DESCRIPTION
**Pull request by kinarashah**
Each system image now has information on which addon template to fetch,
this is calculated on metadata refresh based on range information now
(instead of hardcoded k8s major versions). This will also allow providing
addon templates for minor versions.

Issue: 
https://github.com/rancher/rancher/issues/21714

Vendor PRs:
https://github.com/rancher/rke/pull/1530
https://github.com/rancher/kontainer-driver-metadata/pull/19

### CODE REVIEW

**Comment by superseb** [comment]
With the changes on templates and having the ability to mark ranges now, I would prefer it to have all templates bound to a version (range), and default to error. This forces you to make sure all versions we cover have a template, and it will break when you add a new k8s version but not the templates. Also, the more variations are introduced, the more it becomes difficult to see what default gets applied to.

**Comment by kinarashah** [comment]
> With the changes on templates and having the ability to mark ranges now, I would prefer it to have all templates bound to a version (range), and default to error. This forces you to make sure all versions we cover have a template, and it will break when you add a new k8s version but not the templates. Also, the more variations are introduced, the more it becomes difficult to see what default gets applied to.

https://github.com/rancher/rke/issues/1320

File: pkg/settings/setting.go [hunk 1]
```diff
@@ -146,8 +146,8 @@ func NewSetting(name, def string) Setting {
 
 func getMetadataURL() string {
 	data := map[string]interface{}{
-		"url":    "https://github.com/rancher/kontainer-driver-metadata.git",
-		"branch": "master",
+		"url":    "https://github.com/kinarashah/kontainer-driver-metadata.git",
+		"branch": "tee",
```
Comment (prachidamle) [inline pkg/settings/setting.go]: this needs to be changed back?

File: pkg/controllers/management/kontainerdrivermetadata/data_getter.go [hunk 2]
```diff
@@ -45,41 +44,21 @@ func GetRKEWindowsSystemImages(k8sVersion string, sysImageLister v3.RKEK8sWindow
...
+	addon, err := addonLister.Get(namespace.GlobalNamespace, addonName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return "", err
 		}
-		if addon.Labels[sendRKELabel] == "false" {
-			return "", nil
+		addon, err = addons.GetNamespaced(namespace.GlobalNamespace, addonName, metav1.GetOptions{})
+		if !errors.IsNotFound(err) {
+			return "", err
```
Comment (prachidamle) [inline pkg/controllers/management/kontainerdrivermetadata/data_getter.go]: We need to handle if err == errors.IsNotFound? Else we will hit panic below

File: pkg/controllers/management/clusterprovisioner/rke_store.go [hunk 3]
```diff
@@ -1,48 +1,63 @@
...
 
 func (a *rkeStore) GetAddonTemplates(k8sVersion string) map[string]interface{} {
 	data := map[string]interface{}{}
-	for _, addonName := range addonsList {
-		template, err := kd.GetRKEAddonTemplate(k8sVersion, addonName, a.AddonLister, a.Addons)
+	sysImage, err := a.SystemImagesLister.Get(namespace.GlobalNamespace, k8sVersion)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			logrus.Errorf("getAddonTemplates: error finding system image for %s %v", k8sVersion, err)
+			return data
```
Comment (prachidamle) [inline pkg/controllers/management/clusterprovisioner/rke_store.go]: Don't we want to return the error?

Comment (kinarashah) [reply pkg/controllers/management/clusterprovisioner/rke_store.go]: this is getting called from kontainer engine to get data, in this case, we just log the error and move on with the assumption that either rke will have the template or rke will fail with proper msg

File: pkg/controllers/management/kontainerdrivermetadata/data_getter.go (hunk 2 above)
Comment (kinarashah) [reply pkg/controllers/management/kontainerdrivermetadata/data_getter.go]: updated to return all errs if second Get fails

**Comment by prachidamle** [comment]
LGTM

**Pull request merged**